icu4c/source/test/cintltst/cloctst.c

```diff
@@ -6665,6 +6667,155 @@ static void test_unicode_define(const char *namech, char ch, const char *nameu,
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length should set "
+              "status to U_STRING_NOT_TERMINATED_WARNING but got %d %s. "
+              "Returns %s\n",
+              locale, inLocale, status, myErrorName(status), cbuff);
+  }
+  if (buff[length-1] == sentinal1) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length does not "
+              "change memory in the end of buffer while it should. Returns "
+              "'%s'\n",
+              locale, inLocale, cbuff);
+  }
+  if (buff[length] != sentinal2) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length change "
+              "memory beyond maxResultSize to %x. Returns '%s'\n",
+              locale, inLocale, buff[length], cbuff);
+  }
+  if (buff[preflight_length - 1] == 0x0000) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length null "
+              "terminated while it should not. Return '%s'\n",
+              cbuff, locale, inLocale);
+  }
+
+  // 3. Test when we only set the maxResultSize to preflight_length-1.
+  // Set sentinal1 in the buff[preflight_length-1] to check it will not be
+  // replaced with display name.
+  buff[preflight_length-1] = sentinal1;
+  // Set sentinal2 in the buff[preflight_length] to check it won't be replaced
+  // by null.
+  buff[preflight_length] = sentinal2;
+  status = U_ZERO_ERROR;
+  length = uloc_getDisplayName(
+      locale, inLocale, buff, preflight_length - 1, &status);
+
+  convert_status = U_ZERO_ERROR;
+  // Convert the actual result to UTF8 for debugging output.
+  u_strToUTF8(cbuff, 256, &outlength, buff, length-1, &convert_status);
+
+  if (length != preflight_length) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length-1 return "
+              "length %d different from preflight length %d. Returns '%s'\n",
+              locale, inLocale, length, preflight_length, cbuff);
+  }
+  if (U_BUFFER_OVERFLOW_ERROR != status) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length-1 should "
+              "set status to U_BUFFER_OVERFLOW_ERROR but got %d %s. "
+              "Returns %s\n",
+              locale, inLocale, status, myErrorName(status), cbuff);
+  }
+  if (buff[length-1] != sentinal1) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length-1 should "
+              "not change memory in beyond the maxResultSize. Returns '%s'\n",
+              locale, inLocale, cbuff);
+  }
+  if (buff[length] != sentinal2) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length-1 change "
+              "memory beyond maxResultSize to %x. Returns '%s'\n",
+              locale, inLocale, buff[length], cbuff);
+  }
+  if (buff[preflight_length - 2] == 0x0000) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length-1 null "
+              "terminated while it should not. Return '%s'\n",
+              locale, inLocale, cbuff);
+  }
+}
+
+static void Test21157CorrectTerminating(void) {
```

Comment:
FYI (really, no need to change)

I find it more convenient to write unit tests in C++ (intltest). As long as we have full *coverage* of C API in cintltst, making sure everything can be used/called from plain C, we can test further functionality from C++.